### PR TITLE
chore: upgrade SQLAlchemy to 1.4

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

Testing upgrading SQLAlchemy to 1.4, based on https://docs.sqlalchemy.org/en/14/changelog/migration_14.html.

The biggest change is that the `URL` object is no longer mutable, so I had to update the methods that were modifying it inplace, making them return a modified copy instead.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by craig-rueda — APPROVED** [review]
LGTM

File: tests/integration_tests/config_tests.py [hunk 1]
```diff
@@ -137,6 +137,9 @@ def test_python_date_format_by_column_name(self):
         self.assertEqual(dms_col.python_date_format, "invalid")
 
     def test_expression_by_column_name(self):
```
Comment (betodealmeida) [inline tests/integration_tests/config_tests.py:139]: This tests only fails in SQLite, for some reason. I decided to skip it.

Comment (betodealmeida) [reply tests/integration_tests/config_tests.py]: I was not able to reproduce the failing SQLite tests no matter what — I even tried https://github.com/nektos/act and running the containers manually. So I rewrote them as proper unit tests. :-P

**Review by villebro — APPROVED** [review]
Wonderful PR! ❤️ Just a few minor non-blocking comments, but this is a great step forward, as 1.3 is starting to become less and less supported (For instance, I set out to bump Pandas from 1.3 to 1.4 in #20083, but Pandas 1.4 no longer supports SQLAlchemy 1.3). Once this is merged I think we should do a bump pass on our SQLA connectors, as many of the versions we've pinned to may not yet support SQLA 1.4.

**Comment by rxm7706** [comment]
Any timeline on when this  might be merged ?

**Comment by betodealmeida** [comment]
> Any timeline on when this might be merged ?

I need to spend some time figuring out why some SQLite tests are failing. I'll give it another try this week.

**Comment by EugeneTorap** [comment]
Hi @betodealmeida, any updates?

**Comment by betodealmeida** [comment]
> Hi @betodealmeida, any updates?

I'm merging this today. :)

**Comment by EugeneTorap** [comment]
@villebro @betodealmeida Do we plan upgrade pandas to 1.4.3, pyarrow to 8.0.0, numpy to 1.23.1?

**Comment by betodealmeida** [comment]
> @villebro @betodealmeida Do we plan to upgrade pandas to 1.4.3, pyarrow to 8.0.0, numpy to 1.23.1?

We usually only upgrade dependencies if there's a need or a security issue, since it can always break things. Do you have any need for newer versions of this libraries?

**Comment by daniel-sali** [comment]
Hi @betodealmeida,

Do you have a timeline for when this will be released? I'm asking because we use Superset with Snowflake, and snowflake-sqlalchemy requires sqlalchemy 1.4 or newer, and older versions limit the snowflake-connector-python Python driver version to an unsupported old version. 
Unfortunately, this change is not in the latest 1.5 and 2.0 Superset releases yet :-(

Thanks,
Daniel

**Comment by EugeneTorap** [comment]
Hello @daniel-sali !
sqlalchemy 1.4 will be in 2.0.1 release (this or next week)
Now we have 2.0.1rc2 - https://github.com/apache/superset/blob/2.0.1rc2/CHANGELOG.md

**Comment by daniel-sali** [comment]
Thank you @EugeneTorap for the quick reply, best of luck with the new release, will give it a go once it's out.

**Comment by daniel-sali** [comment]
Hi @EugeneTorap, I don't see the change in the 2.0.1 RC2. I'm checking setup.py for the sqlalchemy dependency version, and it's still specified as < 1.4

**Comment by EugeneTorap** [comment]
@betodealmeida maybe you know it?

**Comment by dkhwangbo** [comment]
@betodealmeida @EugeneTorap 
Hey. This PR is not merged in 2.0.1. Any timeline to release new version this pr merged?

**Comment by EugeneTorap** [comment]
Hi @dkhwangbo!
It will be merged in 2.1.0 version (in 2-3 months)

**Comment by cwegener** [comment]
@dkhwangbo 2.1.0.rc1 has this change included. https://github.com/apache/superset/discussions/23164

**Pull request merged**